plugins/discourse-ai/lib/automation/llm_tagger.rb

```diff
@@ -0,0 +1,166 @@
+# frozen_string_literal: true
+
+module DiscourseAi
+  module Automation
+    module LlmTagger
+      def self.handle(
+        post:,
+        tagger_persona_id:,
+        tag_mode: "manual",
+        available_tags: [],
+        confidence_threshold:,
+        max_tags:,
+        max_post_tokens:,
+        automation: nil
+      )
+        # Skip if manual mode but no tags provided
+        return if tag_mode == "manual" && available_tags.empty?
+
+        tagger_persona = AiPersona.find(tagger_persona_id)
+        model_id = tagger_persona.default_llm_id || SiteSetting.ai_default_llm_model
+        return if model_id.blank?
+        model = LlmModel.find(model_id)
+
+        if tag_mode == "manual"
+          # Manual mode: provide specific tag list in prompt
+          available_tags_text = "Available tags: #{available_tags.join(", ")}\n\n"
+          input = "#{available_tags_text}Post to analyze:\ntitle: #{post.topic.title}\n#{post.raw}"
+        else
+          # Discover mode: instruct AI to use list_tags tool
+          input =
+            "Post to analyze:\ntitle: #{post.topic.title}\n#{post.raw}\n\nUse the list_tags tool to see available tags, then suggest appropriate ones."
+        end
+
+        if max_post_tokens.present?
+          input =
+            model.tokenizer_class.truncate(
+              input,
+              max_post_tokens,
+              strict: SiteSetting.ai_strict_token_counting,
+            )
+        end
+
+        if post.upload_ids.present?
+          input = [input]
+          input.concat(post.upload_ids.map { |upload_id| { upload_id: upload_id } })
+        end
+
+        bot =
+          DiscourseAi::Personas::Bot.as(
+            Discourse.system_user,
+            persona: tagger_persona.class_instance.new,
+            model: model,
+          )
+
+        persona_response_format = tagger_persona.response_format
+
+        bot_ctx =
+          DiscourseAi::Personas::BotContext.new(
+            user: Discourse.system_user,
+            skip_tool_details: true,
+            feature_name: "llm_tagger",
+            messages: [{ type: :user, content: input }],
+          )
+
+        llm_args = {
+          feature_context: {
+            automation_id: automation&.id,
+            automation_name: automation&.name,
+          },
+        }
+
+        if persona_response_format.present? && !Rails.env.test?
```

Comment:
Do we need `!Rails.env.test?`. I don't think this will work without a response format, right?